internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,705 @@
+                    DisplayList {
+                        values: &bounds,
+                        delim: ", "
+                    },
+                    self.expr,
+                )
+            }
+            None => write!(
+                f,
+                "forall [{}] {}",
+                DisplayList {
+                    values: &self.free,
+                    delim: ", "
+                },
+                self.expr
+            ),
+        }
+    }
+}
+
+// Kind represents a class or family of types
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Kind {
+    Addable,
+    Subtractable,
+    Divisible,
+    Comparable,
+    Nullable,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Kind::Addable => f.write_str("Addable"),
+            Kind::Subtractable => f.write_str("Subtractable"),
+            Kind::Divisible => f.write_str("Divisible"),
+            Kind::Comparable => f.write_str("Comparable"),
+            Kind::Nullable => f.write_str("Nullable"),
+        }
+    }
+}
+
+// MonoType represents a specific named type
+#[derive(Debug, Clone, PartialEq)]
+pub enum MonoType {
+    Bool,
+    Int,
+    Uint,
+    Float,
+    Str,
+    Duration,
+    Time,
+    Regexp,
+    Var(Tvar),
+    Arr(Box<Array>),
+    Row(Box<Row>),
+    Fun(Box<Function>),
+}
+
+impl fmt::Display for MonoType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MonoType::Bool => f.write_str("bool"),
+            MonoType::Int => f.write_str("int"),
+            MonoType::Uint => f.write_str("uint"),
+            MonoType::Float => f.write_str("float"),
+            MonoType::Str => f.write_str("string"),
+            MonoType::Duration => f.write_str("duration"),
+            MonoType::Time => f.write_str("time"),
+            MonoType::Regexp => f.write_str("regexp"),
+            MonoType::Var(var) => var.fmt(f),
+            MonoType::Arr(arr) => arr.fmt(f),
+            MonoType::Row(obj) => obj.fmt(f),
+            MonoType::Fun(fun) => fun.fmt(f),
+        }
+    }
+}
+
+// Tvar stands for type variable.
+// A type variable holds an unknown type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Tvar(i64);
+
+impl fmt::Display for Tvar {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "t{}", self.0)
+    }
+}
+
+// Array is a homogeneous list type
+#[derive(Debug, Clone, PartialEq)]
+pub struct Array(MonoType);
+
+impl fmt::Display for Array {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[{}]", self.0)
+    }
+}
+
+// Row is an extensible record type.
+//
+// A row is either Empty meaning it has no properties,
+// or it is an extension of a row.
+//
+// A row may extend what is referred to as a row
+// variable. A row variable is a type variable that
+// represents an unknown record type.
+//
+#[derive(Debug, Clone)]
+pub enum Row {
```

Comment:
I guess this is what `Object` is now in our Go implementation?

Or maybe not Object in the sense of `values.Object`, because that's a value and not a type.
Do we have an equivalent in Go?

Reply:
Yes you're right, this is the replacement for what is currently the Object type in Go. It looks a little different than the Go type as it is designed to be extensible. But it serves the same purpose.